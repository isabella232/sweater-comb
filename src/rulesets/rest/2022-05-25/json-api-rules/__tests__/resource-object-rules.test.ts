import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { RuleRunner, TestHelpers } from "@useoptic/rulesets-base";
import { context } from "../../__tests__/fixtures";
import { resourceObjectRules } from "../resource-object-rules";

const baseJson = TestHelpers.createEmptySpec();

describe("resource object rules", () => {
  describe("valid GET responses", () => {
    test("passes when status code 200 has the correct JSON body", () => {
      const afterJson = {
        ...baseJson,
        paths: {
          "/api/example": {
            get: {
              responses: {
                "200": {
                  description: "",
                  content: {
                    "application/vnd.api+json": {
                      schema: {
                        type: "object",
                        properties: {
                          data: {
                            type: "object",
                            properties: {
                              id: {
                                type: "string",
                                format: "uuid",
                              },
                              type: {
                                type: "string",
                              },
                            },
                          },
                          links: {
                            properties: {
                              self: {
                                type: "string",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      } as OpenAPIV3.Document;

      const ruleRunner = new RuleRunner([resourceObjectRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);
      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(true);
      expect(results).toMatchSnapshot();
    });
  });

  describe("valid post shapes", () => {
    test("passes when status code 201 has the correct headers and body", () => {
      const afterJson = {
        ...baseJson,
        paths: {
          "/api/example": {
            post: {
              responses: {
                "201": {
                  description: "",
                  headers: {
                    location: {},
                  },
                  content: {
                    "application/vnd.api+json": {
                      schema: {
                        type: "object",
                        properties: {
                          data: {
                            type: "object",
                            properties: {
                              id: {
                                type: "string",
                                format: "uuid",
                              },
                              type: {
                                type: "string",
                              },
                            },
                          },

                          links: {
                            properties: {
                              self: {
                                type: "string",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      } as OpenAPIV3.Document;

      const ruleRunner = new RuleRunner([resourceObjectRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);
      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(true);
      expect(results).toMatchSnapshot();
    });
  });

  describe("valid patch shapes", () => {
    test("passes when status code 200 has the correct body", () => {
      const afterJson = {
        ...baseJson,
        paths: {
          "/api/example": {
            patch: {
              responses: {
                "200": {
                  description: "",
                  content: {
                    "application/vnd.api+json": {
                      schema: {
                        type: "object",
                        properties: {
                          data: {
                            type: "object",
                            properties: {
                              id: {
                                type: "string",
                                format: "uuid",
                              },
                              type: {
                                type: "string",
                              },
                            },
                          },
                          jsonapi: {
                            type: "string",
                          },
                          links: {
                            properties: {
                              self: {
                                type: "string",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      } as OpenAPIV3.Document;

      const ruleRunner = new RuleRunner([resourceObjectRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);
      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(true);
      expect(results).toMatchSnapshot();
    });
  });

  describe("valid delete shapes", () => {
    test("passes when status code 200 has the correct body", () => {
      const afterJson = {
        ...baseJson,
        paths: {
          "/api/example": {
            delete: {
              responses: {
                "200": {
                  description: "",
                  content: {
                    "application/vnd.api+json": {
                      schema: {
                        type: "object",
                        properties: {
                          jsonapi: {
                            type: "string",
                          },
                          meta: {
                            type: "string",
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      } as OpenAPIV3.Document;

      const ruleRunner = new RuleRunner([resourceObjectRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);
      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(true);
      expect(results).toMatchSnapshot();
    });
  });

  test("fails when content is specified for 204 status codes", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/example": {
          patch: {
            responses: {
              "204": {
                description: "",
                content: {
                  "application/vnd.api+json": {
                    schema: {
                      type: "string",
                    },
                  },
                },
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([resourceObjectRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });
});
