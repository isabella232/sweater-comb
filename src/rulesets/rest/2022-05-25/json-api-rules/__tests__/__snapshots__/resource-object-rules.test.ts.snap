// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resource object rules fails when content is specified for 204 status codes 1`] = `
Array [
  Object {
    "change": Object {
      "added": Object {
        "description": "",
        "statusCode": "204",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "204",
          },
          "method": "patch",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "patch",
          "responses",
          "204",
        ],
        "jsonPath": "/paths/~1api~1example/patch/responses/204",
        "kind": "response",
      },
    },
    "condition": "not include content for 204 status codes",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": "expected response to not have content",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "empty content for 204 status codes",
    "passed": false,
    "received": undefined,
    "where": "added response status code: 204 in operation: PATCH /api/example",
  },
]
`;

exports[`resource object rules valid GET responses passes when status code 200 has the correct JSON body 1`] = `
Array [
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "get",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/get/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "include JSON:API data property for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: GET /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "get",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/get/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "self links",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: GET /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "get",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/get/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid get / post response data schema",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: GET /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "description": "",
        "statusCode": "200",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1api~1example/get/responses/200",
        "kind": "response",
      },
    },
    "condition": "include content for 2xx status codes other than 204",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "content for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 in operation: GET /api/example",
  },
]
`;

exports[`resource object rules valid delete shapes passes when status code 200 has the correct body 1`] = `
Array [
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "delete",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "delete",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/delete/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "include JSON:API type property for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: DELETE /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "delete",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "delete",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/delete/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid patch response data schema",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: DELETE /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "description": "",
        "statusCode": "200",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "delete",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "delete",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1api~1example/delete/responses/200",
        "kind": "response",
      },
    },
    "condition": "include content for 2xx status codes other than 204",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "content for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 in operation: DELETE /api/example",
  },
]
`;

exports[`resource object rules valid patch shapes passes when status code 200 has the correct body 1`] = `
Array [
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "patch",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "patch",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/patch/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "response data for patch",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: PATCH /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "patch",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "patch",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/patch/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "include JSON:API type property for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: PATCH /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "patch",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "patch",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/patch/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "self links",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: PATCH /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "patch",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "patch",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/patch/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid patch response data schema",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 with content-type: application/vnd.api+json  in operation: PATCH /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "description": "",
        "statusCode": "200",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "patch",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "patch",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1api~1example/patch/responses/200",
        "kind": "response",
      },
    },
    "condition": "include content for 2xx status codes other than 204",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "content for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 200 in operation: PATCH /api/example",
  },
]
`;

exports[`resource object rules valid post shapes passes when status code 201 has the correct headers and body 1`] = `
Array [
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "201",
          },
          "method": "post",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "post",
          "responses",
          "201",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/post/responses/201/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "include JSON:API data property for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 201 with content-type: application/vnd.api+json  in operation: POST /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "201",
          },
          "method": "post",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "post",
          "responses",
          "201",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/post/responses/201/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "self links",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 201 with content-type: application/vnd.api+json  in operation: POST /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "contentType": "application/vnd.api+json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "201",
          },
          "method": "post",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "post",
          "responses",
          "201",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1example/post/responses/201/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": "match expected shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "valid get / post response data schema",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 201 with content-type: application/vnd.api+json  in operation: POST /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "description": "",
        "statusCode": "201",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "201",
          },
          "method": "post",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "post",
          "responses",
          "201",
        ],
        "jsonPath": "/paths/~1api~1example/post/responses/201",
        "kind": "response",
      },
    },
    "condition": "include content for 2xx status codes other than 204",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "content for 2xx status codes",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 201 in operation: POST /api/example",
  },
  Object {
    "change": Object {
      "added": Object {
        "description": "",
        "statusCode": "201",
      },
      "changeType": "added",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "201",
          },
          "method": "post",
          "path": "/api/example",
        },
        "conceptualPath": Array [
          "operations",
          "/api/example",
          "post",
          "responses",
          "201",
        ],
        "jsonPath": "/paths/~1api~1example/post/responses/201",
        "kind": "response",
      },
    },
    "condition": "have response header matching shape",
    "docsLink": "https://jsonapi.org/format/#document-resource-objects",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "location header",
    "passed": true,
    "received": undefined,
    "where": "added response status code: 201 in operation: POST /api/example",
  },
]
`;
